runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/getdata/ApplianceGetDataClient.java

```diff
@@ -0,0 +1,253 @@
+      builder.addRequests(computationBuilder);
+    }
+
+    try {
+      Windmill.GetDataResponse response = windmillClient.getData(builder.build());
+      // Dispatch the per-key responses back to the waiting threads.
+      for (Windmill.ComputationGetDataResponse computationResponse : response.getDataList()) {
+        for (Windmill.KeyedGetDataResponse keyResponse : computationResponse.getDataList()) {
+          pendingResponses
+              .get(
+                  WindmillComputationKey.create(
+                      computationResponse.getComputationId(),
+                      keyResponse.getKey(),
+                      keyResponse.getShardingKey()))
+              .set(keyResponse);
+        }
+      }
+    } catch (RuntimeException e) {
+      // Fan the exception out to the reads.
+      for (QueueEntry entry : batch.reads) {
+        entry.response.setException(e);
+      }
+    } finally {
+      synchronized (this) {
+        Preconditions.checkState(activeReadThreads >= 1);
+        if (pendingReadBatches.isEmpty()) {
+          activeReadThreads--;
+        } else {
+          // Notify the thread responsible for issuing the next batch read.
+          ReadBatch startBatch = pendingReadBatches.remove(0);
+          startBatch.startRead.set(true);
+        }
+      }
+    }
+  }
+
+  /**
+   * Adds the entry to a read batch for sending to the windmill server. If a non-null batch is
+   * returned, this thread will be responsible for sending the batch and should wait for the batch
+   * startRead to be notified. If null is returned, the entry was added to a read batch that will be
+   * issued by another thread.
+   */
+  private @Nullable ReadBatch addToReadBatch(QueueEntry entry) {
+    synchronized (this) {
+      ReadBatch batch;
+      if (activeReadThreads < MAX_ACTIVE_READS) {
+        assert (pendingReadBatches.isEmpty());
+        activeReadThreads += 1;
+        // fall through to below synchronized block
+      } else if (pendingReadBatches.isEmpty()
+          || pendingReadBatches.get(pendingReadBatches.size() - 1).reads.size()
+              >= MAX_READS_PER_BATCH) {
+        // This is the first read of a batch, it will be responsible for sending the batch.
+        batch = new ReadBatch();
+        pendingReadBatches.add(batch);
+        batch.reads.add(entry);
+        return batch;
+      } else {
+        // This fits within an existing batch, it will be sent by the first blocking thread in the
+        // batch.
+        pendingReadBatches.get(pendingReadBatches.size() - 1).reads.add(entry);
+        return null;
+      }
+    }
+    ReadBatch batch = new ReadBatch();
+    batch.reads.add(entry);
+    batch.startRead.set(true);
+    return batch;
+  }
+
+  private static final class ReadBatch {
+    ArrayList<QueueEntry> reads = new ArrayList<>();
+    SettableFuture<Boolean> startRead = SettableFuture.create();
```

Comment:
nit: does SettableFuture<Void> work?
we never set to false, we're just using the future to trigger. Void would capture the value doesn't matter better.

Reply:
if we set it to void i think the value would always be null
so also removing       Preconditions.checkState(batch.startRead.get());

Reply:
Thanks!